SO-2959: Properly prevent deletion of released components

nagyo opened the pull request:
- always use one SnomedDeletionPlan per SnomedEditingContext
- allow deleting the last FSN of a concept
- simplified validation of description type concepts
- added unit tests for concepts, relationships and descriptions

nagyo committed:
SO-2959: Properly prevent deletion of released components

- always use one SnomedDeletionPlan per SnomedEditingContext
- allow deleting the last FSN of a concept
- simplified validation of description type concepts
- added unit tests for concepts, relationships and descriptions

https://snowowl.atlassian.net/browse/SO-2959

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java:
```diff
@@ -557,58 +557,67 @@ public void delete(EObject object, boolean force) {
 	}
 	
 	private void delete(Concept concept, boolean force) {
```
With a single `SnomedDeletionPlan` you could remove the component-specific delete methods, keep the `canDelete(...)` methods and move the code that throws the rejection exception at the end of the public `delete(EObject, boolean)` method.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java:
```diff
@@ -619,126 +628,96 @@ public SnomedDeletionPlan canDelete(Concept concept, SnomedDeletionPlan deletion
+		if (isDescriptionType.isPresent()) {
+			
+			boolean hasRelatedDescriptions = SnomedRequests.prepareSearchDescription()
+				.setLimit(1)
```
You can safely limit the result set to zero. The returned collection resource will still have a total number reflecting the total number of matches.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java:
```diff
@@ -619,126 +628,96 @@ public SnomedDeletionPlan canDelete(Concept concept, SnomedDeletionPlan deletion
 		}
 		
-		deletionPlan.markForDeletion(refSetEditingContext.getReferringMembers(relationship));		
+		deletionPlan.markForDeletion(refSetEditingContext.getReferringMembers(relationship));
```
Further performance improvements could be made here by moving the `getReferringMembers` logic to the `preCommit` phase just like we did it with the actual component deletion.

cmark reviewed (approved):
LGTM!

nagyo committed:
SO-2959: Further simplify component deletion

https://snowowl.atlassian.net/browse/SO-2959

nagyo committed:
SO-2959: Simplify component deletion, remove unused method

https://snowowl.atlassian.net/browse/SO-2959

Commit:
SO-2959: Further improvement of deletion by getting all...

... referring members of components in the preCommit phase rather than
getting it component by component.

apeteri commented on snomed/com.b2international.snowowl.snomed.api.rest.tests/src/com/b2international/snowowl/snomed/api/rest/components/SnomedRelationshipApiTest.java:
```diff
@@ -60,6 +64,8 @@
  */
 public class SnomedRelationshipApiTest extends AbstractSnomedApiTest {
 
+	private static final String ROOT_ISA_RELATIONSHIP_ID = "1019504021";
```
The prefix `ROOT_` could be changed to `RELEASED_` or similar across all test cases; it is a bit misleading at the moment, and it doesn't really matter who the "owner" is, as long as the component is released.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedDeletionPlan.java:
```diff
@@ -84,11 +82,8 @@ private void internalMarkForDeletion(final Collection<? extends CDOObject> items
-		descriptionsToUpdate.addAll(Sets.newHashSet(descriptions));
+	@Override
+	public String toString() {
+		return rejectionReasons.size() < 2 ? Iterables.getFirst(rejectionReasons, "") : Joiner.on(", ").join(rejectionReasons);
```
`Joiner` will handle an empty or single-element `Iterable` in the expected manner:

```java
if (parts.hasNext()) {
  appendable.append(toString(parts.next()));
  while (parts.hasNext()) {
      appendable.append(separator);
      appendable.append(toString(parts.next()));
  }
}
```

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java:
```diff
@@ -542,300 +549,114 @@ private Concept getConceptChecked(final CDOID cdoId) {
+				tryDelete(relationship, force);
+				
+				if (deletionPlan.isRejected()) {
+					deletionPlan.addRejectionReason(
```
Each released inbound relationship now results in two rejection messages, the regular "relationship is released" as well as "concept has released inbound relationship" one. Good, bad?

zoliMolnar replied:
I think this is fine because it will make it more clear why it failed when we log the rejection reasons.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java:
```diff
@@ -976,9 +783,24 @@ private Relationship buildDefaultIsARelationship(Concept parentConcept, Concept
-		if (deletionPlan != null) {
-			delete(deletionPlan);
+		
+		if (!deletionPlan.isRejected() && !deletionPlan.isEmpty()) {
```
Shouldn't we throw an exception if execution reaches this point and the deletion plan is rejected? This will silently skip the deletion parts of the commit.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/SnomedEditingContext.java:
```diff
@@ -998,6 +820,180 @@ public void preCommit() {
+				refSetEditingContext.getContents().remove(index);
+			}
+			else if (eObject instanceof SnomedRefSetMember) {
+				// get the refset and remove the member from it's list
```
it's &rarr; its

Commit:
SO-2959: Removed check from toString, if the deletionplan is rejected...

... throw a ConflictException in preCommit phase, renamed constant in
SnomedRelationshipApiTest.

apeteri committed:
SO-2959: Create new instance of the deletion plan after removing items
